### PR TITLE
Catalogue Service - Multi project build approach

### PR DESCRIPTION
**Pull request by msupic**
Refactored the catalogue service to contain the following subprojects:
- app - contains the application code with no cloud specific dependencies or configuration.
- aws - depends on the `app` project and introduces configuration and dependencies that integrate the application with services of AWS
- oci - depends on the `app` project and introduces configuration and dependencies that integrate the application with services of Oracle Cloud

_Note: Still need to update README.md_

### CODE REVIEW

File: deploy/complete/helm-chart/mushop/charts/catalogue/config/atp.init.sql [hunk 1]
```diff
@@ -51,89 +51,6 @@ BEGIN
...
-		SELECT COUNT(*) 
-		INTO tableExists 
-		FROM DBA_TABLES 
-		WHERE owner = '&1'
-		AND table_name = tableName;
-		DBMS_OUTPUT.PUT_LINE ('** Table creationg steps - &_DATE');
-		IF tableExists = 0 THEN
-			DBMS_OUTPUT.PUT_LINE ('Creating Table ' || tableName || '...' );
-			EXECUTE IMMEDIATE 'CREATE TABLE &1..' || tableName || ' (
-				sku VARCHAR2(20) NOT NULL,
```
Comment (msupic) [inline deploy/complete/helm-chart/mushop/charts/catalogue/config/atp.init.sql:68]: Tables are created now by flyway so there is no need to have this code here anymore.

File: src/catalogue/app/src/main/java/catalogue/Application.java [hunk 2]
```diff
@@ -32,6 +33,9 @@
 public class Application {
 
     public static void main(String[] args) {
-        Micronaut.run(Application.class, args);
+        Micronaut.build(args)
+                .mainClass(Application.class)
+                .defaultEnvironments(Environment.DEVELOPMENT)
```
Comment (alvarosanchez) [inline src/catalogue/app/src/main/java/catalogue/Application.java:38]: Why is this needed?

Comment (alvarosanchez) [reply src/catalogue/app/src/main/java/catalogue/Application.java:38]: Ok, I see now that you put the DB configuration in an environment-specific config file.

Defaulting the environment in `Application` is a bad practise. The environment should be defined when running the application. It could be part of the entrypoint of the `Dockerfile` (`java -Dmicronaut.environments=dev..`), but in this case it is much better to define the `MICRONAUT_ENVIRONMENTS` environment variable in the Kubernetes deployment manifest.

Comment (alvarosanchez) [reply src/catalogue/app/src/main/java/catalogue/Application.java:38]: Also, the `dev` name is misleading, since you are using that to deploy to "production". Perhaps `in-memory` or `h2` are better alternatives.

File: src/catalogue/aws/src/main/java/catalogue/AwsApplication.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package catalogue;
+
+import io.micronaut.context.env.Environment;
+import io.micronaut.runtime.Micronaut;
+
+public class AwsApplication {
+    public static void main(String[] args) {
+        Micronaut.build(args)
+                .mainClass(AwsApplication.class)
+                .defaultEnvironments(Environment.AMAZON_EC2)
```
Comment (alvarosanchez) [inline src/catalogue/aws/src/main/java/catalogue/AwsApplication.java:10]: This environment will be deduced automatically when deploying in AWS.

Anyway we will disable deduction by default in Micronaut 4, so you may want to disable it here, and then activate the environment via `MICRONAUT_ENVIRONMENTS` as said in the other comment.

File: src/catalogue/app/src/main/java/catalogue/repositories/CategoryRepository.java [hunk 4]
```diff
@@ -1,10 +1,13 @@
...
 
 import catalogue.model.Category;
+import io.micronaut.data.jdbc.annotation.JdbcRepository;
+import io.micronaut.data.model.query.builder.sql.Dialect;
 import io.micronaut.data.repository.CrudRepository;
 
 import java.util.List;
 
+@JdbcRepository(dialect = Dialect.H2)
 public interface CategoryRepository extends CrudRepository<Category, Long> {
```
Comment (alvarosanchez) [inline src/catalogue/app/src/main/java/catalogue/repositories/CategoryRepository.java:11]: What if we rename it to `H2CategoryRepository` or `InMemoryCategoryRepository`?

File: src/catalogue/app/src/main/java/catalogue/repositories/ProductRepository.java [hunk 5]
```diff
@@ -3,13 +3,16 @@
...
 import io.micronaut.data.model.Pageable;
+import io.micronaut.data.model.query.builder.sql.Dialect;
 import io.micronaut.data.repository.PageableRepository;
 
 import javax.validation.constraints.NotNull;
 import java.util.List;
 import java.util.Optional;
 
+@JdbcRepository(dialect = Dialect.H2)
 public interface ProductRepository extends PageableRepository<Product, String> {
```
Comment (alvarosanchez) [inline src/catalogue/app/src/main/java/catalogue/repositories/ProductRepository.java:16]: What if we rename it to `H2ProductRepository ` or `InMemoryProductRepository `?

File: src/catalogue/oci/src/main/java/catalogue/OciApplication.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package catalogue;
+
+import io.micronaut.context.env.Environment;
+import io.micronaut.runtime.Micronaut;
+
+public class OciApplication {
+    public static void main(String[] args) {
+        Micronaut.build(args)
+                .mainClass(OciApplication.class)
+                .defaultEnvironments(Environment.ORACLE_CLOUD)
```
Comment (alvarosanchez) [inline src/catalogue/oci/src/main/java/catalogue/OciApplication.java:10]: See comment for `AwsApplication`

**Review by alvarosanchez — CHANGES_REQUESTED** [review]
Beyond the inline comments, I would really like to see `catalogue` being a truly multi-project build, with shared build code across the submodules to avoid duplication.

Also, the tests look the same for both modules, so again to avoid duplication, you could have a "tck" module where the tests are defined, and then the specific AWS/OCI ones simply extend from the TCK. See https://github.com/alvarosanchez/object-storage-samples/ for an example.

File: .github/workflows/java-ci.yml [hunk 7]
```diff
@@ -91,7 +91,11 @@ jobs:
           if [ -f "mvnw" ]; then
             ./mvnw deploy -Dpackaging=docker -Djib.from.image=ghcr.io/graalvm/graalvm-ce:java11-21.1.0 -Ddocker.image.suffix=graalvm -Dproject.version=$GITHUB_SHA --no-transfer-progress
           elif [ -f "gradlew" ]; then
-            ./gradlew dockerPush -Pversion=$GITHUB_SHA --no-daemon
+            if [ "${{ matrix.project-path }}" == "catalogue" ]; then
+              ./gradlew :app:dockerPush -Pversion=$GITHUB_SHA --no-daemon
+            else
+              ./gradlew dockerPush -Pversion=$GITHUB_SHA --no-daemon
+            fi
```
[the remark this answers is not in the document]
Comment (melix) [reply .github/workflows/java-ci.yml:98]: If you do `gradle dockerPush`, then it will execute _all_ `dockerPush` tasks in _all_ projects. So if `:app` defines one, it will be called, but if `:aws` _also_ defines one, then it will be pushed too. There's no need to create a top level `dockerPush`. I think something clearer would be to move the top level `src` directory into its own subproject. That's how "idiomatic Gradle" recommends to setup projects, with the TLP being basically an empty shell, and subprojects defining components.

**Comment by msupic** [comment]
@alvarosanchez - I was told to refactor the MuShop micronaut services to be similar to this demo project: https://github.com/graemerocher/multicloud-demo. Almost all above comments are related to code/ideas that were taken from the demo project, so I am not sure should we keep the code as is to be similar to the demo project or to change per comments above.
@graemerocher - please take a look at comments above

**Comment by graemerocher** [comment]
suggestions for improvements by @alvarosanchez aren't bad, I'm not sure if the need to use `MICRONAUT_ENVIRONMENTS` though it seems just as easy to hard code it into the `main` method like already done in the PR

**Comment by alvarosanchez** [comment]
It isn't a big deal anyway. Since this project is a showcase, I thought I'd recommend what I think are best practices.

**Comment by graemerocher** [comment]
yeah we should probably use the same patterns as in GCN

File: src/catalogue/app/src/main/java/catalogue/repositories/ProductRepository.java (hunk 5 above)
Comment (msupic) [reply src/catalogue/app/src/main/java/catalogue/repositories/ProductRepository.java:16]: The same comment as for CategoryRepository

File: src/catalogue/app/src/main/java/catalogue/repositories/CategoryRepository.java (hunk 4 above)
Comment (alvarosanchez) [reply src/catalogue/app/src/main/java/catalogue/repositories/CategoryRepository.java:11]: Maybe leave it as it is

**Comment by msupic** [comment]
@alvarosanchez - I refactored tests per this comment
```
Also, the tests look the same for both modules, so again to avoid duplication, you could have a "tck" module where the tests are defined, and then the specific AWS/OCI ones simply extend from the TCK. See https://github.com/alvarosanchez/object-storage-samples/ for an example.
```
Please let me know if there is something else that needs to be changed.

File: src/catalogue/app/src/main/java/catalogue/Application.java (hunk 2 above)
Comment (msupic) [reply src/catalogue/app/src/main/java/catalogue/Application.java]: @alvarosanchez @graemerocher - Maybe we should use `app` instead of `h2` or something else that is not specific just to `h2` database because I need to setup additional configuration just for that environment.

Currently the `app` subproject has
```
  metrics:
    export:
      prometheus:
        enabled: true
        descriptions: true
        step: PT1M
```
in the `application.yml` file. But that configuration should be in the `application-h2.yml` file since it is specific only for the `app` subproject (should not be used in `aws` and `oci` subprojects).

Also the following needs to be added to the `app` subproject configuration
```
endpoints:  
    prometheus:
        enabled: true
        sensitive: false
        details-visible: ANONYMOUS
```

**Review by alvarosanchez — APPROVED** [review]
